### ISSUE TITLE
Order of input messed up when auto typing

### ISSUE THREAD

**Issue by Arvedui** [jwilm/alacritty#1588]
Which operating system does the issue occur on?
linux

If on linux, are you using X11 or Wayland?
X11

When auto typing with xdotools default delay (14ms) ordering of key strokes is not preserved most of the time and with even lower delays key presses are sometimes even dropped.
Depending on the alacritty version a delay around 30ms seems to be safe. 
The glutin 0.18 PR improves the situation but does not fix it.

I tried this with several other terminals (xterm, termite, kitty) they all receive the simulated key presses without problems even when disabling the delay entirely.

**Comment by jwilm** [comment]
@Arvedui that sounds bad. Would you mind trying to reproduce this with the winit window [example](https://github.com/tomaka/winit/blob/master/examples/window.rs)?

Alternatively, can you provide detailed instructions to reproduce?

**Comment by Arvedui** [comment]
Steps to reproduce:
1. `sleep 1; xdotool type --delay 12  "abcdefghijklmnopqrstuvwxyz"` (sleep is only there to give one enough time to switch the window)
2. switch immediately to an alacritty window.
3. characters should appear in a messed up order. If not decrease delay and try again.

I could not reproduce this with the winit window example based on the terminal output.

I also ran alacritty with `--print-events`, the messed up order is clearly visible in the output as you can see. 

	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 30, state: Pressed, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 30, state: Released, virtual_keycode: Some(A), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 48, state: Pressed, virtual_keycode: Some(B), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 48, state: Released, virtual_keycode: Some(B), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 46, state: Pressed, virtual_keycode: Some(C), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 46, state: Released, virtual_keycode: Some(C), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 32, state: Pressed, virtual_keycode: Some(D), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 32, state: Released, virtual_keycode: Some(D), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 18, state: Pressed, virtual_keycode: Some(E), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 18, state: Released, virtual_keycode: Some(E), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 33, state: Released, virtual_keycode: Some(F), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 33, state: Pressed, virtual_keycode: Some(F), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 34, state: Pressed, virtual_keycode: Some(G), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 34, state: Released, virtual_keycode: Some(G), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 38, state: Pressed, virtual_keycode: Some(L), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 16, state: Released, virtual_keycode: Some(Q), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 16, state: Pressed, virtual_keycode: Some(Q), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 25, state: Released, virtual_keycode: Some(P), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 25, state: Pressed, virtual_keycode: Some(P), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 24, state: Released, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 24, state: Pressed, virtual_keycode: Some(O), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 49, state: Released, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 49, state: Pressed, virtual_keycode: Some(N), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 50, state: Released, virtual_keycode: Some(M), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 50, state: Pressed, virtual_keycode: Some(M), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 38, state: Released, virtual_keycode: Some(L), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 35, state: Pressed, virtual_keycode: Some(H), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 35, state: Released, virtual_keycode: Some(H), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 23, state: Pressed, virtual_keycode: Some(I), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 23, state: Released, virtual_keycode: Some(I), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 36, state: Pressed, virtual_keycode: Some(J), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 36, state: Released, virtual_keycode: Some(J), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 37, state: Pressed, virtual_keycode: Some(K), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 37, state: Released, virtual_keycode: Some(K), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 19, state: Pressed, virtual_keycode: Some(R), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 19, state: Released, virtual_keycode: Some(R), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 31, state: Pressed, virtual_keycode: Some(S), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 31, state: Released, virtual_keycode: Some(S), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 20, state: Pressed, virtual_keycode: Some(T), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 20, state: Released, virtual_keycode: Some(T), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 22, state: Pressed, virtual_keycode: Some(U), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 17, state: Pressed, virtual_keycode: Some(W), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 17, state: Released, virtual_keycode: Some(W), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 45, state: Pressed, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 45, state: Released, virtual_keycode: Some(X), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 44, state: Pressed, virtual_keycode: Some(Y), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 21, state: Released, virtual_keycode: Some(Z), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 21, state: Pressed, virtual_keycode: Some(Z), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }
	glutin event: WindowEvent { window_id: WindowId(X(WindowId(62914565))), event: KeyboardInput { device_id: DeviceId(X(DeviceId(3))), input: KeyboardInput { scancode: 44, state: Released, virtual_keycode: Some(Y), modifiers: ModifiersState { shift: false, ctrl: false, alt: false, logo: false } } } }


This means it is a problem in glutin, right?

**Comment by Arvedui** [comment]
I played around with the winit example a little and tried to mimic alacrittys event loop.
It looks like this now:
```rust
extern crate winit;

use std::{thread, time};

fn main() {
    let mut events_loop = winit::EventsLoop::new();
    let window = winit::Window::new(&events_loop).unwrap();

    loop {
        events_loop.run_forever(|event| {
            println!("forever: {:?}", event);

            winit::ControlFlow::Break
        });

        let sleep_time = time::Duration::from_millis(50);
        thread::sleep(sleep_time);

        events_loop.poll_events(|event| {
            println!("poll: {:?}", event);
            thread::sleep(time::Duration::from_millis(5));
        });
    }
}
```

With this and a xdotool type delay of 30 milliseconds I get … interesting results.
The event order is somewhat reliably completely flipped!
I than looked closer at the messed up character sequences I get in alacritty and it seems that most if not all can be put in their original order just by flipping consecutive characters.
For instance:
```
a b c d e f h g i j k m p o n l q r s v y x w t u z
            g h       l n o p m       v u t w x y
                        m p o n       t u v
                          n o p
```
I take this as a strong hint that what I can produce with winit directly and the problem I experience in alacrity have the same root cause.
Therefore I'd say this is a race condition somewhere in winit or below.

**Comment by jwilm** [comment]
cc @francesca64

**Comment by Arvedui** [comment]
I believe I found the origin of this bug.

I stumbled upon [this](https://gitlab.freedesktop.org/xorg/lib/libx11/issues/34) bug report for libx11, which winit uses for communication with x11.
It only occurs when using a xim, which I do.
I can reproduce this bug in other applications as well and it completely vanishes in all of them (including alacritty) when disabling the xim.
I only noticed this in alacritty because most applications I use use either gtk or qt. Both have their own input module apis that are preferred over xim and are obviously not affected, however when forcing them to use xim they are affected as well.

Sorry for the noise. :)
